### PR TITLE
Implement MDS039 build directive rule

### PR DESCRIPTION
**Pull request by jeduden**
## Summary

Implement the MDS039 lint rule that validates `<?build?>` directives in Markdown files and keeps their generated bodies in sync with recipe templates. This enables declaring build artifacts with configurable recipes defined in `.mdsmith.yml`.

## Key Changes

- **New rule implementation** (`internal/rules/build/rule.go`):
  - Validates `<?build?>` directives with required parameters (`recipe`, `output`)
  - Enforces recipe resolution against user-declared recipes in config
  - Prevents unsafe output paths (rejects `..` path components)
  - Validates required/optional parameters per recipe schema
  - Generates section bodies from recipe `body_template` with `{output}` and `{alt}` placeholders
  - Detects stale bodies and reports them as errors
  - Warns on unknown parameters (non-fatal)

- **Comprehensive test suite** (`internal/rules/build/rule_test.go`):
  - 343 lines of unit tests covering all validation paths
  - Tests for metadata (ID, name, category)
  - Settings parsing and validation
  - Recipe resolution and parameter validation
  - Body generation with custom and default templates
  - Integration tests for the full Check/Fix workflow
  - Edge cases (dotdot paths, unknown recipes, stale bodies)

- **Configuration integration** (`internal/config/build.go`):
  - Inject parsed recipes into both MDS039 (build) and MDS040 (recipe-safety) rules
  - Recipes serialized from config into rule settings

- **Documentation**:
  - Rule documentation (`internal/rules/MDS039-build/README.md`)
  - User guide (`docs/guides/directives/build.md`)
  - Test fixtures with good/bad examples and expected diagnostics

- **Test fixtures** (`internal/rules/MDS039-build/`):
  - Good examples: user-declared recipes, any file extension
  - Bad examples: unknown recipe, missing output, dotdot paths, missing required params, unknown params, stale body
  - Fixed example showing corrected stale body

- **Integration**:
  - Register rule in `cmd/mdsmith/main.go` and integration tests
  - Update plan status to complete (✅)
  - Update rule index and documentation index

## Implementation Details

- Uses `gensection.Engine` for directive parsing and body regeneration
- Recipes are user-declared only (no built-in recipes); all must be in `build.recipes` config
- Default body template is `[{output}]({output})` when recipe omits `body-template`
- `{alt}` placeholder defaults to `"{recipe} output: {output}"`
- Validation is split: hard errors (missing recipe/output, unknown recipe, missing required params) prevent body generation; unknown params produce warnings alongside valid bodies
- Path safety check rejects any segment that is exactly `..` (handles both `/` and `\` separators)

https://claude.ai/code/session_01TBfp8tSTzhz5m5Kw9iPV87

### CODE REVIEW

File: internal/rules/build/rule.go [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+	}
+
+	for _, req := range schema.Required {
+		if _, ok := params[req]; !ok {
+			return []lint.Diagnostic{gensection.MakeDiag(
+				r.RuleID(), r.RuleName(), filePath, line,
+				fmt.Sprintf("build directive recipe %q: missing required parameter %q", recipeName, req),
+			)}
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go:204]: Fixed in dbf7832. `validateHard` now uses `strings.TrimSpace(v) == ""` to treat empty/whitespace required param values as missing.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/guides/directives/build.md [hunk 2]
```diff
@@ -0,0 +1,106 @@
+---
+summary: >-
+  How to use the build directive to declare artifact outputs, keep
+  generated bodies in sync, and configure user-declared recipes.
+---
```
[the remark this answers is not in the document]
Comment (jeduden) [reply docs/guides/directives/build.md:6]: Fixed in dbf7832. Added `title: Build directive` to the front matter and regenerated `docs/guides/index.md`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/build/rule.go [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+			`build directive missing required "output" parameter`,
+		)}
+	}
+
+	if hasDotDotSegment(output) {
+		return []lint.Diagnostic{gensection.MakeDiag(
+			r.RuleID(), r.RuleName(), filePath, line,
+			fmt.Sprintf(`build directive "output" contains ".." path component: %q`, output),
+		)}
+	}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go:187]: Fixed in dbf7832. `validateHard` now calls `filepath.IsAbs(output)` and rejects absolute paths before the `..` check.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/index.md [hunk 4]
```diff
@@ -57,6 +57,7 @@ row: "| [{id}]({filename}) | `{name}` | {status} | {description} |"
 | [MDS036](MDS036-max-section-length/README.md)                 | `max-section-length`                 | ready     | Section length must not exceed per-level or per-heading limits.                                                     |
 | [MDS037](MDS037-duplicated-content/README.md)                 | `duplicated-content`                 | ready     | Paragraphs should not repeat verbatim across Markdown files.                                                        |
 | [MDS038](MDS038-toc/README.md)                                | `toc`                                | ready     | Keep toc generated heading lists in sync with document headings.                                                    |
+| [MDS039](MDS039-build/README.md)                              | `build`                              | ready     | Validate <?build?> directive parameters and keep the section body in sync with the recipe's rendered body_template. |
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/index.md]: Fixed in 351a42d. Both `<?build?>` and `body_template` are now backtick-wrapped in the index row.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/build/rule.go [hunk 5]
```diff
@@ -0,0 +1,350 @@
...
+
+	schema, _ := r.resolveRecipe(recipeName)
+	tmpl := schema.BodyTemplate
+	if tmpl == "" {
+		tmpl = defaultBodyTemplate
+	}
+
+	alt := fmt.Sprintf("%s output: %s", recipeName, output)
+	body := strings.ReplaceAll(tmpl, "{alt}", alt)
+	body = strings.ReplaceAll(body, "{output}", output)
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in 351a42d. Switched to `strings.NewReplacer("{alt}", alt, "{output}", output).Replace(tmpl)` for single-pass substitution.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/build/rule.go [hunk 6]
```diff
@@ -0,0 +1,350 @@
...
+			`build directive missing required "output" parameter`,
+		)}
+	}
+
+	if filepath.IsAbs(output) {
+		return []lint.Diagnostic{gensection.MakeDiag(
+			r.RuleID(), r.RuleName(), filePath, line,
+			fmt.Sprintf(`build directive "output" must be a relative path: %q`, output),
+		)}
+	}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in 351a42d. The condition is now `filepath.IsAbs(output) || filepath.VolumeName(output) != ""`, which catches `C:out.png`-style paths on Windows.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/build/rule.go [hunk 7]
```diff
@@ -0,0 +1,350 @@
...
+	for name, rawRecipe := range rawMap {
+		rm, ok := rawRecipe.(map[string]any)
+		if !ok {
+			return nil, fmt.Errorf("recipe %q must be a map, got %T", name, rawRecipe)
+		}
+		schema := recipeSchema{}
+		if bt, ok := rm["body-template"]; ok {
+			if s, ok := bt.(string); ok {
+				schema.BodyTemplate = s
+			}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in 351a42d. `parseRecipesSettings` now returns an error when `body-template` is present but not a string.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/index.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/index.md]: Fixed in b384481. Updated the `description` in `MDS039-build/README.md` to use backticks around `<?build?>` and `` `body-template` ``, then regenerated `internal/rules/index.md`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/MDS039-build/README.md [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+name: build
+status: ready
+description: >-
+  Validate <?build?> directive parameters and keep the section body
+  in sync with the recipe's rendered body_template.
+---
+# MDS039: build
+
+Validate <?build?> directive parameters and keep the section body
+in sync with the recipe's rendered body_template.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS039-build/README.md]: Fixed in b384481. Both the frontmatter `description` and the opening body paragraph now use `` `<?build?>` `` and `` `body-template` ``.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: docs/guides/directives/build.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+# Build directive
+
+The `<?build?>` directive declares a build artifact — a file
+produced by a recipe configured in `build.recipes`. `mdsmith fix`
+renders the section body from the recipe's `body_template` and
+keeps it up to date. No external tool runs at lint time.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply docs/guides/directives/build.md:12]: Fixed in b384481. Changed `body_template` → `` `body-template` `` in the guide prose to match the YAML config key.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/build/rule_test.go [hunk 10]
```diff
@@ -0,0 +1,468 @@
...
+		{"a/b/c.png", false},
+		{"out.png", false},
+		{"./out.png", false},
+		{"..", true},
+		{"a/b/..c.png", false},
+	}
+	for _, c := range cases {
+		assert.Equal(t, c.want, hasDotDotSegment(c.path), "path=%q", c.path)
+	}
+}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule_test.go:375]: Fixed in b384481. Added `` `a\..\b.png` `` (true) and `` `a\b\c.png` `` (false) cases to `TestHasDotDotSegment`.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/rules/build/rule.go [hunk 11]
```diff
@@ -0,0 +1,351 @@
...
+
+// recipeSchema holds the param schema and body template for a recipe.
+type recipeSchema struct {
+	Required     []string
+	Optional     []string
+	BodyTemplate string
+}
+
+// defaultBodyTemplate is the fallback body_template for recipes that omit body-template.
+const defaultBodyTemplate = "[{output}]({output})"
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in d246602.

File: internal/rules/MDS039-build/README.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+3. **Required params present** — params listed as required by the
+   recipe schema must all be supplied.
+4. **No unknown params** (warning) — params not in the recipe's
+   `required` or `optional` lists produce a warning.
+5. **Body in sync** — the section body must equal the rendered
+   `body_template`; MDS039 reports `generated section is out of date`
+   when it diverges.
+
+`mdsmith fix` rewrites the body using the rendered `body_template`.
+No external tool is executed.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS039-build/README.md:31]: Fixed in d246602.

File: internal/rules/MDS039-build/README.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
...
+| `recipe` | yes      | Recipe name declared in `build.recipes`                         |
+| `output` | yes      | Artifact path relative to the Markdown file; no `..` components |
+
+`output` accepts any file extension; MDS039 applies no extension
+filter.
+
+## Generated body
+
+Each recipe has a `body_template` rendered by `mdsmith fix`. Two
+placeholders are available:
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS039-build/README.md:58]: Fixed in d246602.

File: docs/guides/directives/build.md [hunk 14]
```diff
@@ -0,0 +1,107 @@
...
+?>
+![render output: docs/diagram.png](docs/diagram.png)
+<?/build?>
+```
+
+## Generated body
+
+`mdsmith fix` renders the section body from the recipe's
+`body_template`. Two placeholders are available:
+
```
[the remark this answers is not in the document]
Comment (jeduden) [reply docs/guides/directives/build.md:77]: Fixed in d246602.

File: internal/rules/MDS039-build/README.md [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+
+## What it detects
+
+MDS039 validates each `<?build?>` directive in a Markdown file:
+
+1. **`recipe` resolves** — the recipe name must be declared in
+   `build.recipes` in `.mdsmith.yml`.
+2. **`output` is present and safe** — `output` is required and
+   must not contain `..` path components.
+3. **Required params present** — params listed as required by the
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/MDS039-build/README.md:22]: Fixed in d246602.

File: plan/101_build-directive-mds039.md [hunk 16]
```diff
@@ -123,87 +87,78 @@ in `build.recipes.NAME.body_template` (plan 100).
...
-   be supplied by the directive.
+3. **Required params present** — `params.required`
+   entries from the recipe schema must all be
+   supplied by the directive.
 4. **No unknown params** — params not in the recipe's
    `required` or `optional` lists produce a warning.
 5. **Body in sync** — the section body must equal the
    rendered `body_template`. MDS039 reports
    `stale-section` when it diverges; `Fix` rewrites
    the body using `gensection.Engine`.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply plan/101_build-directive-mds039.md]: Fixed in d246602.

File: docs/guides/directives/build.md [hunk 17]
```diff
@@ -0,0 +1,107 @@
...
+MDS039 validates `<?build?>` directives and reports:
+
+- **Error** when `recipe` is missing or not declared in `build.recipes`
+- **Error** when `output` is missing or contains `..` components
+- **Error** when a required param for the recipe is absent
+- **Warning** when a param is not in the recipe's `required` or
+  `optional` lists
+- **Error** (`generated section is out of date`) when the body
+  diverges from the rendered `body_template`
```
[the remark this answers is not in the document]
Comment (jeduden) [reply docs/guides/directives/build.md]: Fixed in d246602.

File: docs/guides/directives/build.md [hunk 18]
```diff
@@ -0,0 +1,107 @@
...
+
+When `body-template` is omitted from the recipe declaration, the
+default `[{output}]({output})` is used.
+
+## Rule MDS039
+
+MDS039 validates `<?build?>` directives and reports:
+
+- **Error** when `recipe` is missing or not declared in `build.recipes`
+- **Error** when `output` is missing or contains `..` components
```
[the remark this answers is not in the document]
Comment (jeduden) [reply docs/guides/directives/build.md]: Fixed in d246602.

File: internal/rules/build/rule.go [hunk 19]
```diff
@@ -0,0 +1,351 @@
+// Package build implements MDS039, which validates <?build?> directive
+// parameters and keeps the body in sync with the recipe's body_template.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in d246602.

File: docs/guides/directives/build.md [hunk 20]
```diff
@@ -0,0 +1,107 @@
...
+it cannot introduce a new recipe inline.
+
+```yaml
+build:
+  recipes:
+    render:
+      command: "myrenderer {source} -o {output}"
+      body-template: "![{alt}]({output})"
+      params:
+        required: [source]
```
[the remark this answers is not in the document]
Comment (jeduden) [reply docs/guides/directives/build.md:52]: Fixed in bb37fc0.

File: internal/config/build_test.go [hunk 21]
```diff
@@ -276,6 +276,70 @@ func TestInjectBuildConfig_OverwritesExistingSettings(t *testing.T) {
...
 
+func TestInjectBuildConfig_BuildRule(t *testing.T) {
+	cfg := &Config{
+		Build: BuildConfig{
+			Recipes: map[string]RecipeCfg{
+				"render": {
+					Command:      "myrenderer {source} -o {output}",
+					BodyTemplate: "![{alt}]({output})",
+					Params: ParamCfg{
+						Required: []string{"source"},
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/config/build_test.go:312]: Fixed in bb37fc0.

File: plan/101_build-directive-mds039.md [hunk 22]
```diff
@@ -123,87 +87,78 @@ in `build.recipes.NAME.body_template` (plan 100).
...
 
-- [ ] `<?build recipe:screenshot url:... output:...?>`
-      body is regenerated on `mdsmith fix`
-- [ ] `<?build recipe:vhs input:demo.tape output:demo.gif?>`
-      body is regenerated on `mdsmith fix`
-- [ ] MDS039 reports `stale-section` when the body
+- [x] `<?build?>` body is regenerated on `mdsmith fix`
+      using the recipe's `body_template`
+- [x] MDS039 reports `stale-section` when the body
       diverges from the rendered `body_template`
```
[the remark this answers is not in the document]
Comment (jeduden) [reply plan/101_build-directive-mds039.md]: Fixed in bb37fc0.

File: plan/101_build-directive-mds039.md [hunk 23]
```diff
@@ -123,87 +87,78 @@ in `build.recipes.NAME.body_template` (plan 100).
...
   on conflict; artifact bytes are not regenerated.
 
 ## Tasks
 
-1. Implement the `<?build?>` directive in
+1. [x] Implement the `<?build?>` directive in
    `internal/rules/build/` using `gensection.Engine`.
    Register as MDS039, category `meta`. `Generate`
    renders `body_template` only; it never calls a
    builder or touches the filesystem.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply plan/101_build-directive-mds039.md:123]: Fixed in bb37fc0.

File: internal/rules/build/rule.go [hunk 24]
```diff
@@ -0,0 +1,351 @@
...
+	}
+	return r.engine
+}
+
+// Check implements rule.Rule.
+// It validates each <?build?> directive and reports stale-section when
+// the rendered body differs from the expected body_template output.
+// Unknown params are reported as warnings, which the gensection engine
+// cannot emit alongside a stale-section check, so Check is implemented
+// manually.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in bb37fc0.

File: internal/rules/build/rule.go [hunk 25]
```diff
@@ -0,0 +1,351 @@
...
+			RuleID:   r.RuleID(),
+			RuleName: r.RuleName(),
+			Severity: lint.Warning,
+			Message:  fmt.Sprintf("build directive recipe %q: unknown parameter %q", recipeName, k),
+		})
+	}
+	return diags
+}
+
+// generateBody renders the recipe's body_template using the directive params.
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in bb37fc0.

File: plan/101_build-directive-mds039.md [hunk 26]
```diff
@@ -123,87 +87,78 @@ in `build.recipes.NAME.body_template` (plan 100).
...
-- [ ] `output` accepts any file extension; no
+- [x] `output` accepts any file extension; no
       extension filter is applied
-- [ ] The rendered `body_template` uses `{alt}`
-      defaulting to `"{recipe} output: {output}"`,
-      satisfying MDS032
-- [ ] A user-declared recipe's `body_template` from
+- [x] The rendered `body_template` uses `{alt}`
+      defaulting to `"{recipe} output: {output}"`
+- [x] A user-declared recipe's `body_template` from
```
[the remark this answers is not in the document]
Comment (jeduden) [reply plan/101_build-directive-mds039.md]: Fixed in b0c5ff2.

File: internal/rules/build/rule.go [hunk 27]
```diff
@@ -0,0 +1,351 @@
...
+			`build directive missing required "output" parameter`,
+		)}
+	}
+
+	if filepath.IsAbs(output) || filepath.VolumeName(output) != "" {
+		return []lint.Diagnostic{gensection.MakeDiag(
+			r.RuleID(), r.RuleName(), filePath, line,
+			fmt.Sprintf(`build directive "output" must be a relative path: %q`, output),
+		)}
+	}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in 2a22858. Added `strings.HasPrefix(output, `\\`)` to catch Windows-rooted (`\tmp\out.png`) and UNC (`\\server\share\out.png`) paths on all platforms, plus a unit test for both cases.

File: internal/config/build.go [hunk 28]
```diff
@@ -121,31 +121,37 @@ func validateCommandPlaceholders(recipeName, command string, allowed map[string]
...
-// .mdsmith.yml; it is set in the config-path setting so MDS040 can
-// report diagnostics against the
-// right file.
+// InjectBuildConfig copies cfg.Build.Recipes into the recipe-safety and
+// build rule settings. It is called after config loading in main so rules
+// receive their inputs through the normal ApplySettings path. cfgPath is
+// the path to the loaded .mdsmith.yml; it is set in the config-path
+// setting so MDS040 can report diagnostics against the right file.
 func InjectBuildConfig(cfg *Config, cfgPath string) {
 	if cfg == nil || len(cfg.Build.Recipes) == 0 {
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/config/build.go]: Fixed in 33c4047. Removed the `len(cfg.Build.Recipes) == 0` early-exit so the injection (with an empty map when there are no recipes) always runs, overwriting any user-supplied recipe settings in enabled rules.

File: internal/rules/build/rule.go [hunk 29]
```diff
@@ -0,0 +1,351 @@
...
+			r.RuleID(), r.RuleName(), filePath, line,
+			`build directive missing required "output" parameter`,
+		)}
+	}
+
+	if filepath.IsAbs(output) || filepath.VolumeName(output) != "" || strings.HasPrefix(output, `\`) {
+		return []lint.Diagnostic{gensection.MakeDiag(
+			r.RuleID(), r.RuleName(), filePath, line,
+			fmt.Sprintf(`build directive "output" must be a relative path: %q`, output),
+		)}
```
[the remark this answers is not in the document]
Comment (jeduden) [reply internal/rules/build/rule.go]: Fixed in d35bb4e. Replaced `filepath.VolumeName` with an explicit `hasDriveLetter` helper that checks whether `p[0]` is ASCII a–z/A–Z and `p[1] == ':'` — fully OS-independent. Also added `TestValidateHard_WindowsDriveLetter` covering `C:\\out.png`, `C:out.png`, and `d:file.txt`.

**Comment by jeduden** [comment]
🔍 **Merge Queue** — bisecting

A larger batch failed CI. Bisection is isolating the culprit: this run tests up to **2 of 3** candidate PRs on [`merge-queue/batch-bisect-223-1777815378`](https://github.com/jeduden/mdsmith/tree/merge-queue/batch-bisect-223-1777815378). [View current bisect CI run](https://github.com/jeduden/mdsmith/actions/runs/25280642275).

**Next:** No action needed — you'll be notified when the culprit is isolated or this PR merges.

**Comment by jeduden** [comment]
🔍 **Merge Queue** — bisecting

A larger batch failed CI. Bisection is isolating the culprit: this run tests up to **1 of 1** candidate PRs on [`merge-queue/batch-bisect-223-1777839014`](https://github.com/jeduden/mdsmith/tree/merge-queue/batch-bisect-223-1777839014). [View current bisect CI run](https://github.com/jeduden/mdsmith/actions/runs/25289562937).

**Next:** No action needed — you'll be notified when the culprit is isolated or this PR merges.

**Comment by jeduden** [comment]
✅ **Merge Queue** — merged

This PR landed on `main` via commit [`c013640`](https://github.com/jeduden/mdsmith/commit/c0136406c4c46d5f80ec05673c1d54a8f44550d6). [CI run that validated the merge](https://github.com/jeduden/mdsmith/actions/runs/25289562937).

**Next:** Done — nothing more to do here.
